lib/wallet/src/Cardano/Wallet/Transaction.hs

```diff
@@ -497,6 +502,28 @@ emptyWitnessCount = WitnessCount
     , bootstrap = 0
     }
 
+-- WitnessCount context is needed to differentiate verification keys present
+-- in native scripts.
+-- In shelley wallets they could be present due to only policy verification key.
+-- In multisig wallet they could stem from payment, policy and delegation roles,
+-- and as minting/burning and delegation support comes will be extended in additional
+-- data attached in SharedWalletCtx to differentiate that.
+-- WitnessCount is needed only during or after signing, in other phases it is not used.
+data WitnessCountCtx =
+    ShelleyWalletCtx KeyHash | SharedWalletCtx | AnyWitnessCountCtx
+    deriving (Eq, Generic, Show)
+    deriving anyclass NFData
+
+toKeyRole :: WitnessCountCtx -> ByteString -> KeyRole
+toKeyRole witCtx key = case witCtx of
+    ShelleyWalletCtx (KeyHash _ mypolicykey) ->
+        if key == mypolicykey then
+            Policy
+        else
+            Unknown
+    SharedWalletCtx -> Payment
```

Comment:
the decision is to be made based on what a given wallet understands. So if we have shelley wallet it knows its policy verification key. Then if it encounters script with some verification key then there are two situations: one it recognized its policy key and verification key could be translated into KeyHash with Policy role. Otherwise it is trnaslated into KeyHash with Unknown role. For shared wallet we WILL have more options. Right now we assume that all verification keys are assuming Payment role. In the future we will need to differentiate between Policy (when minting will be supported), Delegation (when staking will be supported) and Unknown when veification key does not belong to any known verification keys

Reply:
Ok, but one could then think of it as the HRP being ultimately based off nothing more than: 1) the derivation path to the key + 2) whether that derivation path is "valid" given the wallet type, correct? E.g. `Shelley` wallet can never identify `addr_shared_vkh` keys in native scripts. And for no wallet we let the HRP be influenced by what the script actually unlocks in the tx? This seems sane to me, but wanted to confirm.

Reply:
in the end for shelley only OPTIONALLY minting/burning via scripts are possible.So if it recognizes ver key payload in a script (of course from rootK one can derive all branches, also H1855 used for policy ver key), then it can encapsulate this payload and construct KeyHash (data KeyHash = KeyHash Role ByteString). If it does not recognize it then it is Unknown. That's it. For shared wallet there are more options as ALL credentials could be constructed from script containing verification keys. In that case we will check script template and also if in a given script there is verification key that belongs to us (and for payment also within pool gap). If we detect at least one verification key of ours then we know other verification keys are our cosigners, so we can also also ascribe them Payment. The complete handle of these matters will be possible only as we support more functionalities in shared wallets (minting, staking, ...). But we could fully handle Shelley wallets case and this PR does it